### PR TITLE
nixos/pipewire: add configPackages options

### PR DESCRIPTION
**Pull request by hcsch**
## Description of changes

Added `configPackages` options to both `services.pipewire` and `services.pipewire.wireplumber`. Packages specified in these options have their relevant files (e.g. `share/pipewire/pipewire.conf.d/*`) combined in store path generated by `pkgs.buildEnv` and linked to `/etc/`.
Pipewire `extraConfig` options also get added to this store path in the appropriate locations.

Added `extraLv2Packages` options to the options of both services for packages providing lv2 plugins required by [filter chains](https://docs.pipewire.org/page_module_filter_chain.html) the user set up in their pipewire/wireplumber config files. These get added to the `LV2_PATH` of the service they were specified for. In addition, packages in `configPackages` can specify `passthru.requiredLv2Packages` to have them included in the path as well.

In the case of wireplumber, added config and lv2 packages get added for pipewire as well, since config packages for wireplumber are likely to have config files for pipewire as well.

For now my main use case for this is [asahi-audio](https://github.com/AsahiLinux/asahi-audio) (my setup for that is not yet clean enough to upstream), but I'm sure that as time goes on, other packages will pop up to configure filter chains or similar things. With these options, adding them to nixpkgs should be easy and using them should be easy for users as well.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

I've tested these changes locally with my yet to be upstreamed asahi-audio package and the config files are put in the right places, and lv2 packages are found by wireplumber/pipewire. The existing nixos default config files for pipewire are also present.

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by hcsch** [comment]
Should I document the attribute in `passthru` somewhere for discoverabiltiy? If so, where is the most appropriate place for that?

**Comment by hcsch** [comment]
Forgot some mentions of the not yet upstreamed asahi-audio package in the option docs/examples. Those have been removed with the force-push

**Comment by hcsch** [comment]
@tpwrules If I'm right, you set up asahi-audio for https://github.com/tpwrules/nixos-apple-silicon. I'd be happy to hear if this looks useful to you :)

**Comment by tpwrules** [comment]
@K900 we had discussed something along these lines on Matrix, what do you think? I don't have any opinion yet.

**Comment by hcsch** [comment]
I moved some definitions around in the pipewire module to make the git diff a bit more readable. Adding a `let ... in` in the `config` definition and having to change the indentation accordingly kinda made the diff a bit messy. Hopefully this way it's nicer to review

**Comment by K900** [comment]
Do we really need separate options for LV2 plugins on Wireplumber and Pipewire? I'm not even sure if both need to be aware of it.

**Comment by hcsch** [comment]
You're totally right. I think I just got a bit confused about what requires what to work when I wrote this. I was getting some errors in wireplumber about not being able to find the lv2 plugins, or something along those lines, but all docs (that I found) described this as a feature of pipewire, not wireplumber. And I kinda remember getting some errors from pipewire too, but that might've been due to other issue earlier when I was writing these changes, when I was having some other issues as well.
Anyways, I've taken out the LV2 config options / environment variables from the pipewire module and tested the setup again on my nixos (with nixpkgs at nixos-unstable) and it seemed to worked just fine.

File: nixos/modules/services/desktops/pipewire/wireplumber.nix
Comment (K900) [inline nixos/modules/services/desktops/pipewire/wireplumber.nix:59]: Not a big fan of the name here. Maybe rename the variable and also the file?

**Review by K900 — APPROVED** [review]
Overall looks pretty good

File: nixos/modules/services/desktops/pipewire/wireplumber.nix
Comment (hcsch) [reply nixos/modules/services/desktops/pipewire/wireplumber.nix]: I've changed the name of this config file / config package variable to `pw-not-for-audio`/`pwNotForAudio`. Hopefully that is a more descriptive name. Just "nixos" was indeed a somewhat unhelpful name for the file/variable. Thanks for pointing that out.

In case you're not happy with the new name just yet, I'd appreciate some hint at what you want from the name / perhaps a suggestion.

File: nixos/modules/services/desktops/pipewire/pipewire.nix
Comment (K900) [inline nixos/modules/services/desktops/pipewire/pipewire.nix:233]: Nit: it's not actually `.properties` I think? Unless Pipewire also accepts that? But it's usually .conf or .json

File: nixos/modules/services/desktops/pipewire/wireplumber.nix
Comment (K900) [inline nixos/modules/services/desktops/pipewire/wireplumber.nix:31]: Nit: WirePlumber (capitalization is already all over the place but still no point making it worse)

File: nixos/modules/services/desktops/pipewire/wireplumber.nix
Comment (K900) [inline nixos/modules/services/desktops/pipewire/wireplumber.nix:44]: Nit: LV2 (again capitalization)

File: nixos/modules/services/desktops/pipewire/wireplumber.nix
Comment (hcsch) [reply nixos/modules/services/desktops/pipewire/wireplumber.nix:31]: Do you think having an additional commit to fix the pre-existing capitalisation typos would be good? Otherwise I'll just fix the ones I made and rebase. Sorry by the way, for the descriptions having turned out a bit messy

Comment (K900) [reply nixos/modules/services/desktops/pipewire/wireplumber.nix:31]: If you want to do it, go ahead

**Comment by K900** [comment]
OK let's send it.

**Comment by philiptaron** [comment]
This broke my [sound.nix](https://github.com/philiptaron/nix-config/blob/lex/sound.nix#L60-L62) configuration.

```
ln: failed to create symbolic link '/nix/store/518frv5hpih882nkqq3ap4hk369bxmry-etc/etc/pipewire/pipewire.conf.d/99-input-denoising.conf': Permission denied
```

The fix was to switch to the declarative syntax, which was worth the breakage. See the referenced commit.

**Comment by tpwrules** [comment]
Does this mean there's going to be incompatibility with the stable branch? `environment.etc` being broken is not great.

**Comment by K900** [comment]
`environment.etc` isn't broken, what's broken is declaring a symlink under a path that's already a symlink, e.g. `/etc/pipewire/foo` when `/etc/pipewire` is a symlink to `configPackages`.

**Comment by tpwrules** [comment]
I have not looked carefully but is there a workaround? I'd like to still be able to use `environment.etc` to set up pipewire and such until this becomes the stable branch.

**Comment by K900** [comment]
You can keep doing that, we're not backporting this to stable.

**Comment by philiptaron** [comment]
> `environment.etc` isn't broken, what's broken is declaring a symlink under a path that's already a symlink, e.g. `/etc/pipewire/foo` when `/etc/pipewire` is a symlink to `configPackages`.

Yeah. The link between the two wasn't exactly clear; what I did to diagnose was:

1. `ls -la <parent path that failed>` (in this case, `ls -la /nix/store/518frv5hpih882nkqq3ap4hk369bxmry-etc/etc/pipewire/`)
2. `nix derivation show <symlink target>`
3. Search [NixOS unstable options](https://search.nixos.org/options) for `pipewire.extraConfig` since that was in the derivation name.
4. Click the source link to go to the GitHub page
5. Switch to blame mode
6. Scroll down to find these recently merged changed.

I do wonder what we can do to avoid making a little mess on 23.11 -> 24.05 upgrade for stable folks. Part of the larger "how does upgrade happen" discussion...

**Comment by tpwrules** [comment]
> You can keep doing that, we're not backporting this to stable.

I don't see how it's possible, I don't see a way to get the files out of the new setting and put them back into the old. In any case it turns out I have to break stable compatibility anyway so whatever. Would have been nice to be more careful; I could have reviewed more thoroughly too.

**Comment by LunNova** [comment]
What's the right approach for a downstream pipewire module that wants to work against nixos-unstable and stable?

**Comment by K900** [comment]
Probably check for the presence of the option and fall back if necessary. I've tried to figure out a way to make the migration story better here, but couldn't find a way that didn't involve rewriting most of the /etc module...

**Comment by hcsch** [comment]
Sorry for causing some breakage there. I've also thought a bit about how this could be prevented, but the only thing that came to mind would have been some overlayfs (or other union fs) type addition to `environment.etc`. I suppose we should maybe add an assert to the pipewire module that checks for `environment.etc` entries for PipeWire and warns about them, suggesting what to do instead now (kinda forgot about creating one in the process of this PR). I'll open up another PR for that later today

**Comment by SuperSandro2000** [comment]
Had the same problem and approach as @philiptaron 

```
etc> lrwxrwxrwx 1 nobody nogroup 96 Jan  1  1970 /nix/store/mgkri4plra7x8q10804xvqq3lajq7mmy-etc/etc/pipewire/pipewire.conf.d -> /nix/store/mswjy74qarhfhwl057jp1p3h8mhfag6s-pipewire-extra-config/share/pipewire/pipewire.conf.d
etc> ln: failed to create symbolic link '/nix/store/mgkri4plra7x8q10804xvqq3lajq7mmy-etc/etc/pipewire/pipewire.conf.d/zeroconf.conf': Permission denied
```

This definitely needs improvement before stable, it is very hard to find the root cause of this without advanced NixOS knowledge.

**Comment by SuperSandro2000** [comment]
Also I kinda miss the old way where I could just use text instead of having to create a package with a file first.

**Comment by hcsch** [comment]
For pipewire you can use the `extraConfig` option to provide config as a nix attrset (or whatever else will satisfy `json.type`). For wireplumber I'll likely add an `extraConfig` style option as well exactly because using config packages manually isn't very ergonomic. I'm currently debugging an assertion for pipewire and wireplumber that'll tell you this as-needed

**Comment by tpwrules** [comment]
pkgs.writeText is not that bad to use but I think improvements would definitely be appreciated.

**Comment by hcsch** [comment]
I've created #291946 to add assertions for the new breakage to get better error messages:
```
error:
Failed assertions:
- Using `environment.etc."pipewire<...>"` directly has been deprecated in 24.05. Use `services.pipewire.extraConfig` or `services.piepwire.configPackages` instead.
```

**Comment by andrevmatos** [comment]
This PR seems to have broken `systemWide = true;`. No device is detected when the option is set, while setting it to false makes audio work again. No relevant logs from system's `pipewire.service`, but `wireplumber.service` is giving some errors I didn't notice on previous sessions:
```
wireplumber[33398]: [3:19:05.834753832] [33398]  INFO IPAManager ipa_manager.cpp:143 libcamera is not installed. Adding '/nix/store/src/ipa' to the IPA search path
wireplumber[33398]: [3:19:05.835801344] [33398]  INFO Camera camera_manager.cpp:284 libcamera v0.2.0
systemd[1]: Started Multimedia Service Session Manager.
wireplumber[33398]: <WpPortalPermissionStorePlugin:0x1302300> Failed to call Lookup: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name is not activatable
wireplumber[33398]: <WpPortalPermissionStorePlugin:0x1302300> Failed to call Lookup: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name is not activatable
```

File: nixos/modules/services/desktops/pipewire/wireplumber.nix
Comment (K900) [inline nixos/modules/services/desktops/pipewire/wireplumber.nix:59]: The paths are wrong here.

**Comment by andrevmatos** [comment]
It'd be good to fix before that, as `systemWide` PipeWire is currently broken for any DE, not only plasma6. I can make a PR.

**Comment by K900** [comment]
I want to merge Plasma in the next few hours, FWIW. If you want to make the PR, go ahead.

File: nixos/modules/services/desktops/pipewire/wireplumber.nix
Comment (hcsch) [reply nixos/modules/services/desktops/pipewire/wireplumber.nix:59]: Ah frigg. This really did end up a bit messier than it had to, sorry. I'll create a follow up PR as soon as I can today

**Comment by andrevmatos** [comment]
Done: #292090 
Tested locally and working fine. Thank you

**Comment by hcsch** [comment]
@K900 and the others here: It seems WirePlumber 0.5 changes the wireplumber config location to `/etc/pipewire/wireplumber.conf`/`/etc/pipewire/wireplumber.conf.d` (see https://pipewire.pages.freedesktop.org/wireplumber/daemon/configuration/locations.html#config-locations). https://github.com/NixOS/nixpkgs/pull/278760 seems related as the pre-release for 0.5, with which this should likely cause issues due to this PR.
Given the now seeming abundance of issues with this PR. What do you all think about reverting the PR ASAP to make things run smoothly again for everyone (although the is the risk of re-breaking the config of people who have already started switching to `configPackages`)? With the PR reverted I could then spend some more time to figure out how to get these changes in more cleanly and how to prepare for WirePlumber 0.5 using a config path in `/etc/pipewire`.
If you think the risk of re-breaking configs is too high, then I'll just make a PR for an `extraConfig` option for WirePlumber today and I'll try and figure out how to make the current setup work for when WirePlumber 0.5 lands and the paths change.

**Comment by K900** [comment]
Wireplumber will likely re-revert to /etc/wireplumber for the final 0.5, because it's causing a bunch of issues all over.

**Comment by eclairevoyant** [comment]
I'm not sure I even understood the original motivation for this PR, as the PR description explains what the changes are, rather than _why_.

What was the intention behind switching to configPackages?

Also, I find it extremely questionable that a massive change affecting anyone who uses wayland (plus all X users who use pw) was merged without feedback or testing, and new PRs continue to get merged in this fashion.

I will note that there are zero functionality tests for pw or wp at the moment.

**Comment by K900** [comment]
The reason is to allow easily adding packages that come with their own config files, like hardware specific DSP chains or effects software. The primary consumers for this I know of are Asahi and Jovian, both shipping custom DSP configurations for their target hardware. Also, this change will not affect most Pipewire users, as it generally doesn't require custom configuration, and people who do want to reconfigure it probably know what they're doing.

**Comment by eclairevoyant** [comment]
>  Also, this change will not affect most Pipewire users

If this was really a non-issue, we wouldn't be getting so many reports of it in the Discord...

Beginners will often copy-paste snippets of pw config to get basic functionality working, and this PR (which I again mention has no test cases) will break config for all of these users who definitely are not among those who "know what they're doing".
Even if they know what they are doing with pw that does not necessitate that they know what they are doing on the NixOS side to be aware of packaging (e.g. `pkgs.writeText`) to use `configPackages` directly.
I myself had some issues translating the "pw JSON" (a bespoke format) into compliant JSON to work with `extraConfig` - I don't think even that option solves the breakage issue.

I really cannot get behind this logic of "people who use NixOS must know what they're doing, so let's break things for everyone without feedback, or even being aware of who is affected".

**Comment by eclairevoyant** [comment]
I realised I wasn't clear about what I would want to achieve in the previous message, so here's what I would request:

* Revert all the PRs related to this topic
* Draft a unified PR containing all the changes from the previous PRs (i.e. un-reverting)
* Add, at the very least: release notes mentioning the breaking change, and extensive test cases to identify possible regressions
* Get more than one approval before merging breaking changes with large impact

**Comment by K900** [comment]
I don't think reverting the change is the way to go, as people who have already updated will now have to roll back those changes. What we may want to do is actually fix the `environment.etc` module to handle situations like this correctly, and then downgrade the assertion to a warning. This will allow both existing configuration and new-style configuration to coexist.

**Comment by tpwrules** [comment]
I would be okay with a revert from the Asahi side, it looks simple to check for the existence of the new option and have the module work both ways.

**Comment by eclairevoyant** [comment]
> I don't think reverting the change is the way to go, as people who have already updated will now have to roll back those changes.

Yes, that's correct. But you originally dismissed the breakage for those same users as "they know what they're doing". Meanwhile there's a much larger set of users who haven't updated yet or run into this breakage; I see a much greater benefit to prevent breakage for those users rather than the small subset of users who did in fact figure out how to fix the breakage already.

> What we may want to do is actually fix the `environment.etc` module to handle situations like this

That's a totally valid approach, but that's a change with even more impact, that will require even more scrutiny. In the meantime the PW and WP modules will be broken.

**Comment by hcsch** [comment]
I'd be very happy to have these changes reverted for now, to then spend some time cleaning them up and writing NixOS tests (and perhaps working on fixing `environment.etc`) and later getting these changes merged again with more eyes on it. I'd imagine that reverting now would cause less additional breakage than leaving this as-is.
As @K900 mentioned, the *why* for me was mostly packages that provide pre-configured filter-chains / scripts managing nodes for specific hardware like asahi-audio @eclairevoyant.

**Comment by RANKSHANK** [comment]
Just a heads up this causes a silent error instead of triggering the assert message added in #291946 so you end up with something along the lines of
```error: builder for '/nix/store/z5gbgnp0bs6vnmdv9f91r6ydcfqkh8q8-etc.drv' failed with exit code 1;
       last 1 log lines:
> ln: failed to create symbolic link '/nix/store/x5n02zhb72mz73pzq7lhk59lqx7h01mk-etc/etc/pipewire/pipewire.d': Permission denied```
I wrote a test flake that just enables pipewire and creates a file at /etc/pipewire and that successfully triggers the assert message so I'm not sure what could be causing it to be skipped

**Comment by tpwrules** [comment]
> What's the right approach for a downstream pipewire module that wants to work against nixos-unstable and stable?

https://github.com/tpwrules/nixos-apple-silicon/commit/2baeec1705a47f3b158d8beac79059ccd02ee34c was my approach for what it's worth

**Comment by K900** [comment]
I still think it is too late to revert this, and with https://github.com/NixOS/nixpkgs/pull/291946 providing a better error message, it should hopefully be less confusing for people who are just now upgrading. I'm going to look into the `environment.etc` situation, but probably not right now, so if someone wants to work on that, feel free to snipe.

**Comment by hcsch** [comment]
I'll continue working on the PR to add `extraConfig` / `extraLuaConfig` / `extraScripts` to the WirePlumber module then. If you have some time to spare to take a look at #292115. I'd appreciate comments if you find anything that you'd like changed or are not sure about. The PR is currently still marked as a draft. I'll be posting progress as I make it, and after some clean-up of the commits I'll mark it ready to review. Currently I have a sketch of the options (including hopefully helpful descriptions also describing caveats due to how WirePlumber currently reads config files), but no code just yet to create a config package from these options.

**Comment by budimanjojo** [comment]
@hcsch Can you also update the doc here: https://nixos.wiki/wiki/PipeWire#Bluetooth_Configuration

Or at least tell me what should I do to migrate that chunk of configuration to this change?
Thanks before!

**Comment by Chobbes** [comment]
Yeah, I think it would be really helpful to have an example of how to update existing configurations when an option is deprecated. I know what file I want added to my wireplumber config, but I don't know how to add that with the new changes because it's not just a renaming of an option.

Also, I could be mistaken, but I think the error message should say `services.pipewire.wireplumber.configPackages` instead of `services.wireplumber.configPackages`?

**Comment by hcsch** [comment]
@budimanjojo I'd be very happy to edit the wiki article for the new options. Do you think it would be good to edit it for the current state (`configPackages` only), maybe with a warning that `extraConfig` is coming and edit it again once #292115 lands?

@Chobbes Currently you have to create a nix store path/package which contains your WirePlumber config at `<store-path>/share/wireplumber/<...>`:
```nix
services.pipewire.wireplumber.configPackages = [
  (pkgs.writeTextDir "share/wireplumber/main.lua.d/12-your-config-name-here.lua" ''
    -- Here you can put your Lua config
  '');
]
```
or if you have a lua file next to your NixOS config:
```nix
services.pipewire.wireplumber.configPackages = [
  (pkgs.writeTextDir
    "share/wireplumber/main.lua.d/12-your-config-name-here.lua"
    (builtins.readFile ./12-your-config-name-here.lua)
  );
]
```
`configPackages` will pick up everything in `share/wireplumber` of the store paths it is given and link that to `/etc/wireplumber`.

That will stay the same even with #292115, except that will also add `extraConfig`, `extraLuaConfig`, and `extraScripts` options. With those the above would look like such:
```nix
services.pipewire.wireplumber.extraLuaConfig.main."12-your-config-name-here" = ''
  -- Here you can put your Lua config
'';
```
Which will create and link a file `12-your-config-name-here.lua` into `/etc/wireplumber/main.lua.d/` alongside everything else provided by `configPackages` or other configs/scripts provided to `extra<...>` options.

You're right about the option path currently being incorrect btw, that is fixed by #292326, which should reach nixos-unstable in the near future.

**Comment by hcsch** [comment]
@budimanjojo I've updated https://nixos.wiki/wiki/PipeWire. I'd appreciate it if you could take a look and tell me what you think of the changes, if you have some time :)

**Pull request merged**
